native/dlasy2.go

```diff
@@ -0,0 +1,270 @@
+//  TL*X   + sgn*X*TR^T = scale*B,  if tranl == false and tranr == true,
+//  TL^T*X + sgn*X*TR^T = scale*B,  if tranl == true  and tranr == true,
+// where TL is n1×n1, TR is n2×n2, B is n1×n2, and 1 <= n1,n2 <= 2.
+//
+// isgn must be 1 or -1, and n1 and n2 must be 0, 1, or 2, but these conditions
+// are not checked.
+//
+// Dlasy2 returns a scale factor that is chosen less than or equal to 1 to
+// prevent the solution overflowing, the infinity norm of the solution, and an
+// indicator of success. If ok is false, TL and TR have too close eigenvalues,
+// so TL or TR is perturbed to get a non-singular equation.
+//
+// Dlasy2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlasy2(tranl, tranr bool, isgn, n1, n2 int, tl []float64, ldtl int, tr []float64, ldtr int, b []float64, ldb int, x []float64, ldx int) (scale, xnorm float64, ok bool) {
+	ok = true
+	// Quick return if possible.
+	if n1 == 0 || n2 == 0 {
+		return scale, xnorm, ok
+	}
+
+	// Set constants to control overflow.
+	eps := dlamchP
+	smlnum := dlamchS / eps
+	sgn := float64(isgn)
+
+	if n1 == 1 && n2 == 1 {
+		// 1×1 case: TL11*X + sgn*X*TR11 = B11.
+		tau1 := tl[0] + sgn*tr[0]
+		bet := math.Abs(tau1)
+		if bet <= smlnum {
+			tau1 = smlnum
+			bet = smlnum
+			ok = false
+		}
+		scale = 1
+		gam := math.Abs(b[0])
+		if smlnum*gam > bet {
+			scale = 1 / gam
+		}
+		x[0] = b[0] * scale / tau1
+		xnorm = math.Abs(x[0])
+		return scale, xnorm, ok
+	}
+
+	if n1+n2 == 3 {
+		// 1×2 or 2×1 case.
+		var (
+			smin float64
+			tmp  [4]float64 // tmp is used as a 2×2 column-major matrix.
+			btmp [2]float64
+		)
+		if n1 == 1 && n2 == 2 {
+			// 1×2 case: TL11*[X11 X12] + sgn*[X11 X12]*op[TR11 TR12] = [B11 B12].
+			//                                            [TR21 TR22]
+			smin = math.Abs(tl[0])
+			smin = math.Max(smin, math.Max(math.Abs(tr[0]), math.Abs(tr[1])))
+			smin = math.Max(smin, math.Max(math.Abs(tr[ldtr]), math.Abs(tr[ldtr+1])))
+			smin = math.Max(eps*smin, smlnum)
+			tmp[0] = tl[0] + sgn*tr[0]
+			tmp[3] = tl[0] + sgn*tr[ldtr+1]
+			if tranr {
+				tmp[1] = sgn * tr[ldtr]
+				tmp[2] = sgn * tr[1]
+			} else {
+				tmp[1] = sgn * tr[1]
+				tmp[2] = sgn * tr[ldtr]
+			}
+			btmp[0] = b[0]
+			btmp[1] = b[1]
+		} else {
+			// 2×1 case: op[TL11 TL12]*[X11] + sgn*[X11]*TR11 = [B11].
+			//             [TL21 TL22]*[X21]       [X21]        [B21]
+			smin = math.Abs(tr[0])
+			smin = math.Max(smin, math.Max(math.Abs(tl[0]), math.Abs(tl[1])))
+			smin = math.Max(smin, math.Max(math.Abs(tl[ldtl]), math.Abs(tl[ldtl+1])))
+			smin = math.Max(eps*smin, smlnum)
+			tmp[0] = tl[0] + sgn*tr[0]
+			tmp[3] = tl[ldtl+1] + sgn*tr[0]
+			if tranl {
+				tmp[1] = tl[1]
+				tmp[2] = tl[ldtl]
+			} else {
+				tmp[1] = tl[ldtl]
+				tmp[2] = tl[1]
+			}
+			btmp[0] = b[0]
+			btmp[1] = b[ldb]
+		}
+		// Solve 2×2 system using complete pivoting.
+		// Set pivots less than smin to smin.
+		bi := blas64.Implementation()
```

Comment:
`bi` is used non-conditionally after we have reached L60, so maybe declare it there.